storage/innobase/buf/buf0flu.cc

```diff
@@ -389,21 +381,15 @@ void buf_page_write_complete(const IORequest &request, bool error)
   mysql_mutex_assert_not_owner(&buf_pool.mutex);
   mysql_mutex_assert_not_owner(&buf_pool.flush_list_mutex);
 
-  if (request.is_LRU())
+  if (bpage->oldest_modification() == 2)
   {
-    const bool temp= bpage->oldest_modification() == 2;
-    if (!temp && state < buf_page_t::WRITE_FIX_REINIT &&
-        request.node->space->use_doublewrite())
-      buf_dblwr.write_completed();
     /* We must hold buf_pool.mutex while releasing the block, so that
     no other thread can access it before we have freed it. */
     mysql_mutex_lock(&buf_pool.mutex);
-    bpage->write_complete(temp, error);
+    bpage->write_complete(true, error);
```

Comment:
We already tested the condition here in the caller. Conditional branches are bad for performance. The compiler would not even be allowed to optimize away a redundant check, because `buf_page_t::oldest_modification()` is executing `std::atomic::load()`.

Reply:
It makes sense to declare a separate member function to be invoked on pages of temporary tables. It would not even need the parameter `bool error`:
```c++
    if (!error)
    {
      bpage->write_temporary_complete();
      buf_LRU_free_page(bpage, true);
    }
```

Reply:
I thought more about this. In case of a write error, we do need to release the write fix so that a write can be retried later, or the block evicted from the buffer pool after a `DROP TEMPORARY TABLE` or similar. What we can do is to replace the parameter `bool temporary` with `bool persistent` to simplify the start of the function. We should also pass the already read `bpage->state()` value to that function to avoid the redundant reload. (The way we release the io-fix tolerates concurrent changes of the buffer-fix count.)
```c++
inline void buf_page_t::write_complete(bool persistent, bool error, uint32_t state)
{
  ut_ad(!persistent == fsp_is_system_temporary(id().space()));
  if (UNIV_LIKELY(!error))
  {
    ut_d(lsn_t om= oldest_modification());
    ut_ad(om >= 2);
    ut_ad(persistent == (om > 2));
    /* We use release memory order to guarantee that callers of
    oldest_modification_acquire() will observe the block as
    being detached from buf_pool.flush_list, after reading the value 0. */
    oldest_modification_.store(persistent, std::memory_order_release);
  }
```
Also the caller can be simplified: we evict the block only if `!persistent && !error`. Else, we can use the normal code path, simply invoke `bpage->write_complete(persistent, error, state);`. The predicate `fil_space_t::use_doublewrite()` never holds on the temporary tablespace.

Reply:
Addressed in d157a3c0c6557b8db8ce2ff6764e2fad0cf21821. I don’t think that it makes sense for the function to be `inline` anymore.